### PR TITLE
Address feedback from original change to introduce new dependency resolution algorithm

### PR DESCRIPTION
**Pull request by jeffkl**
# Bug

Tracked at https://github.com/NuGet/Client.Engineering/issues/2959

## Description
My original change to introduce a new dependency resolution algorithm was rushed and this change address the following feedback:

* See if we can get rid of the new exception handling path of AggregateException.  This was only needed because some code paths call `.GetAwaiter().GetResult()` instead of `await` but we undid some of that during the final refactoring.  (https://github.com/NuGet/NuGet.Client/pull/5963#discussion_r1713137804)
* Consider moving SharedStringBuilder to NuGet.Versioning instead of having it a "shared" compiled resource (https://github.com/NuGet/NuGet.Client/pull/5963#discussion_r1713110371)
* Use an inline suppression for ProjectRestoreCommand.GetRuntimeGraph() catching a general exception and remove it from GlobalSupressions.cs (https://github.com/NuGet/NuGet.Client/pull/5963#discussion_r1713111541)
* Fix whitespace in ProjectRestoreCommand.DownloadDependenciesAsync()  (https://github.com/NuGet/NuGet.Client/pull/5963#discussion_r1713112453)
* Fix whitespace in RestoreCommand.ExecuteRestoreAsync() (https://github.com/NuGet/NuGet.Client/pull/5963#discussion_r1713138458)
* Consider using `NoAllocEnumerator` or a `for` loop in `RestoreCommand.CreateFrameworkRuntimePairs()` (https://github.com/NuGet/NuGet.Client/pull/5963#discussion_r1713140673)
* Reconsider marking properties in `RestoreTargetGraph` as `internal` and instead use a new constructor or `init` (https://github.com/NuGet/NuGet.Client/pull/5963#discussion_r1713142543)
* Use extension method for getting the TargetFramework from the PackageSpec (https://github.com/NuGet/NuGet.Client/pull/5965/files#r1715638641)
* Use new property from PackageSpec in [PackageSpecReferenceDependencyProvider.cs](https://github.com/NuGet/NuGet.Client/pull/5965/files/fd329745e696406d6d86bfbf865fdbe0ea40b4ef#diff-b7633a3b599a74c8939203cd6ce8907c506beb2fb3d5f4c22cbe81651103d690) (https://github.com/NuGet/NuGet.Client/pull/5965#discussion_r1715668576)

I broke each one out into its own commit if that's eaiser to review.

## PR Checklist

- [x] Meaningful title, helpful description and a linked NuGet/Home issue
- [ ] Added tests
- [ ] Link to an issue or pull request to update docs if this PR changes settings, environment variables, new feature, etc.

### CODE REVIEW

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ProjectRestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/ProjectRestoreCommand.cs]:
```suggestion
            List<Task<DownloadDependencyResolutionResult>> downloadDependencyResolutionTasks = new(packageSpec.TargetFrameworks.Count);
```

Minimal gains, good practice.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ProjectRestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/ProjectRestoreCommand.cs]: Use NoAllocEnumerate. 

Similar as the rest, good practice, minimal gains.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: I'm not a fan of this change. 
readonly values with defaults set outside of the constructor can lead to bugs. 
It's best if the property is set in a single place like the constructor.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:1227]: Can you help me understand why this is being done? 

This is regressing a previous improvement from https://github.com/NuGet/NuGet.Client/commit/52dd383ae0899680f487ff31e9dd63d0033eea32.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: The return type should be a list.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: The size is knowable here. Let's use it.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs:20]: Given that there's so many changes in this file, can we nullable annotate this now?
It'll be super helfpul

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs:25]: I don't see a shipped.txt change here, I'm guessing it's what's failing the build

**Review by nkolev92 — COMMENTED** [review]
Thanks for addressing all this feedback :)

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/ProjectRestoreCommand.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/ProjectRestoreCommand.cs]: I've updated the method to used properly sized lists and NoAllocEnumerate

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: Okay

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:1227]: A previous version of the prototype had bypassed async functionality by accessing the `.Result` property which was causing exceptions to be made into `AggregateExceptions`.  This was causing tests to fail as the logging code is catching stuff like `ProtocolException`.  In this code path, it wouldn't set `failed` to `true` causing unexpected behavior.

In the final version of the implementation, I have switched to using `.GetAwaiter().GetResult()` which doesn't aggregate exceptions so I think this is not needed anymore.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: Done

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: Done

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs:20]: I think we should tackle that in a different change

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs:25]: This didn't change the public API

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs:20]: We tend to take a best effort approach towards nullability when the type is changed a lot, which I'd say in this case is a bunch of changes. 
I'd prefer if it was done in this PR, but I'm not gonna block this PR on that. 
Your call.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs:25]: Oh duh, internal :D my bad.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: The PackageSpec should not be edited in the restore command. It should really be treated as immutable/
If you look at the telemetry I added, it differentiates between property specified and used and legacy algorithm used because the new resolver doesn't support a certain scenario.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: I think it's `packageSpec.TargetFrameworks.Count * (runtimeIds.Count + 1)` because of the framework, null rid combos that always exist.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:1227]: I think I misread this code change. 
There's a FatalProtocolException catch right above it that still exists there. So it's not really regressing anything. 

Resolving.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: Without this, I have to update a few dozen tests because they only set the `CentralPackageTransitivePinningEnabled` property and not the `UseLegacyDependencyResolver` property.  This seemed cleaner to me than having to update all of those tests.

When I get transitive pinning working for the new resolver, I should be able to remove this line.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: Changing the package spec can mess with no-op of related projects.
We probably don't have tests like that, but it's how we've been able to reduce the amount of package spec cloning we do.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: Basically https://github.com/NuGet/NuGet.Client/pull/5930 is what we'll want eventually.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: > Without this, I have to update a few dozen tests because they only set the CentralPackageTransitivePinningEnabled property and not the UseLegacyDependencyResolver property. This seemed cleaner to me than having to update all of those tests.


Can PackageSpecReferenceDependencyProvider repeat the same check as the one here?
That should remove the need for the test code to change.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: I misread it. I thought you only had multiplication

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: Technically yes, it just places the logic in two places which was how it was before so I had changed it to read this property instead.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: So I expect this to be temporary until we fix the transitive pinning correct?

If that's the case I'd rather duplicate some code, than risk breaking no-op and cause more restores.

Comment (zivkan) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: > Changing the package spec can mess with no-op of related projects.

In general, yes. In this case, it's being mutated in `RestoreCommand`'s constructor, which happens before no-op check (ExecuteAsync?). Personally, I don't care if the first restore after a VS/.NET upgrade fails no-op, only that for the same version of NuGet with the same project inputs has working no-op.

I also completely agree that PackageSpec and DGSpec should be immutable, but even if that was already implemented, this code could have done something like:

```cs
var newRestoreMetadata = new RestoreMetadata(request.Project.RestoreMetadata) { UseLegacy = true} ;
var newProject = new Project(request.Project) { RestoreMetadata = newRestoreMetadata };
_request = new RestoreRequest(request) { Project = newProject };
```

Again, this is in the constructor, before the restore actually starts. Generally agree that our code shouldn't do this, we should design Nuget to deserialize the project inputs into a PackageSpec and then never modify or clone/copy it.

I wish that PackageSpecReferenceDependencyProvider was in NuGet.Commands, not NuGet.ProjectModel (which in my mind should only has DTO objects needed for the DGSpec and maybe APIs to (de)serialize it). In which case we could have an internal helper function and call it in two places, but that's not code we have, so either we duplicate business logic (I have no hesitation in duplicate boilerplate code, but business logic is where Don't Repeat Yourself reduces bugs)

But at the same time, it's temporary until the new resolver implements the feature, and it's a simple if statement that needs to be duplicated.  I'm unconvinced that there's a best solution here. I'm just sharing my thoughts, in case anyone cares.  This PR looks good to me, and I'd be happy to approve once this conversation is resolved.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: PackageSpecReferenceDependencyProvider has the PackageSpec so it can check the exact same condition.

Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: I've removed the setting of this for now.

**Review by nkolev92 — APPROVED** [review]

**Pull request merged**
